src/analysis/processing/qgsalgorithmfieldcalculator.cpp

```diff
@@ -0,0 +1,222 @@
+  const QgsField field(
+    fieldDefinitionMap.value( QStringLiteral( "name" ) ).toString(),
+    fieldTypes.value( fieldType ),
+    fieldType,
+    fieldDefinitionMap.value( QStringLiteral( "length" ) ).toInt(),
+    fieldDefinitionMap.value( QStringLiteral( "precision" ) ).toInt()
+  );
+
+  mFields = source->fields();
+  int fieldIdx = mFields.lookupField( field.name() );
+
+  if ( fieldIdx < 0 )
+    mFields.append( field );
+
+  QString dest;
+  std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, dest, mFields, QgsWkbTypes::multiType( source->wkbType() ), source->sourceCrs() ) );
+
+  mFieldIdx = mFields.lookupField( field.name() );
+
+  // prepare expression
+  QString expressionString = parameterAsString( parameters, QStringLiteral( "FORMULA" ), context );
+  mExpressionContext = createExpressionContext( parameters, context, source.get() );
+  mExpression = QgsExpression( expressionString );
+  mDa.setSourceCrs( source->sourceCrs(), context.transformContext() );
+  mDa.setEllipsoid( context.ellipsoid() );
+
+  mExpression.setGeomCalculator( &mDa );
+  mExpression.setDistanceUnits( context.distanceUnit() );
+  mExpression.setAreaUnits( context.areaUnit() );
+
+  if ( mExpression.hasParserError() )
+    throw QgsProcessingException( QObject::tr( "Parser error with formula expression \"%2\": %3" )
+                                  .arg( expressionString, mExpression.parserErrorString() ) );
+
+
+  return true;
+}
+
+QgsFeatureList QgsFieldCalculatorAlgorithm::processFeature( const QgsFeature &feature, QgsProcessingContext &, QgsProcessingFeedback * )
+{
+  mExpression.prepare( &mExpressionContext );
+
+  QgsAttributes attributes( mFields.size() );
+  const QStringList fieldNames = mFields.names();
+  for ( const QString &fieldName : fieldNames )
+  {
+    const int attributeIndex = feature.fieldNameIndex( fieldName );
+
+    if ( attributeIndex >= 0 )
+      attributes[attributeIndex] = feature.attribute( fieldName );
+  }
+
+  if ( mExpression.isValid() )
+  {
+    mExpressionContext.setFeature( feature );
+    mExpressionContext.lastScope()->setVariable( QStringLiteral( "row_number" ), mRowNumber );
+
+    const QVariant value = mExpression.evaluate( &mExpressionContext );
+
+    if ( mExpression.hasEvalError() )
+    {
+      throw QgsProcessingException( QObject::tr( "Evaluation error in expression \"%1\": %2" )
+                                    .arg( mExpression.expression(), mExpression.evalErrorString() ) );
+    }
+
+    attributes[mFieldIdx] = value;
+    attributes.append( value );
+  }
+  else
+  {
+    attributes.append( QVariant() );
+  }
+
+  QgsFeature f = feature;
+  f.setAttributes( attributes );
+  mRowNumber++;
+  return QgsFeatureList() << f;
+}
+
+
+bool QgsFieldCalculatorAlgorithm::checkParameterValues( const QVariantMap &parameters, QgsProcessingContext &context, QString *message ) const
```

Comment:
I think we can remove all this -- it's already handled in a generic way by the code which checks that non-optional parameters are filled in in the toolbox gui

Reply:
true that, thanks for noticing